Driver exec sometimes fails to SIGKILL a task if the SIGINT is ignored

### Nomad version
Nomad v0. 9.1

### Operating system and Environment details
OS: RedHat Enterprise Linux 7.3

### Issue
Driver exec sometimes fails to SIGKILL a task if the SIGINT is ignored

### Reproduction steps
I have created a simple test job made up of the following two scripts which for simplicity I have put in the /usr/bin directory of my test server:

 trap.sh:
```
#!/bin/bash
trap "echo sigint" SIGINT
trap "echo sigterm" SIGTERM

/usr/bin/trap_sleep.sh
```
trap_sleep.sh:
```
#!/bin/bash
while :
do
    sleep 60
done
```
And am running the job with the following JSON payload:
```
{
  "Job": {
    "Constraints": [],
    "Datacenters": [
      "dc1"
    ],
    "ID": "sigtest",
    "Name": "sigtest",
    "Priority": 50,
    "TaskGroups": [
      {
        "Count": 5,
        "EphemeralDisk": {
          "Migrate": false,
          "SizeMB": 500,
          "Sticky": false
        },
        "Name": "group",
        "Update": {
          "HealthyDeadline": 900000000000,
          "MaxParallel": 1,
          "MinHealthyTime": 10000000000,
          "ProgressDeadline": 0
        },
        "Tasks": [
          {
            "Artifacts": [],
            "Config": {
              "command": "/usr/bin/trap.sh",
              "args": []
            },
            "Driver": "exec",
            "Env": {
              "Redeploy": "1"
            },
            "Leader": true,
            "Name": "sigtest",
            "Resources": {
              "CPU": 200,
              "Devices": null,
              "DiskMB": 0,
              "IOPS": 0,
              "MemoryMB": 75,
              "Networks": []
            },
            "Services": []
          }
        ]
      }
    ],
    "Type": "service",
    "Update": {
      "MaxParallel": 1
    }
  }
}
```
I then try to redeploy the job by increasing the value of the Redeploy env variable and resubmitting.

When Nomad needs to stop the running allocations the script traps the default SIGINT signal and I expect Nomad to force the kill with a SIGKILL five seconds later.

Most of the time this works, but occasionally the SIGKILL is never sent. For example this log file is from an allocation that was never killed:
```
2019-05-20T16:25:21.476+0100 [DEBUG] client.driver_mgr.exec.executor: using plugin: alloc_id=06db34f1-a57e-9004-27eb-942f9bf9e479 driver=exec task_name=sigtest version=2
2019-05-21T10:00:10.408+0100 [DEBUG] client.alloc_migrator: waiting for previous alloc to terminate: alloc_id=56043f29-ac4a-8d31-dfce-4e55f3dfd04a previous_alloc=06db34f1-a57e-9004-27eb-942f9bf9e479
2019-05-21T10:00:30.406+0100 [ERROR] client.alloc_runner.task_runner: failed to kill task: alloc_id=06db34f1-a57e-9004-27eb-942f9bf9e479 task=sigtest backoff=5s error="executor Shutdown failed: rpc error: code = Unknown desc = process failed to exit after 15 seconds"
2019-05-21T10:00:55.410+0100 [ERROR] client.alloc_runner.task_runner: failed to kill task: alloc_id=06db34f1-a57e-9004-27eb-942f9bf9e479 task=sigtest backoff=20s error="executor Shutdown failed: rpc error: code = Unknown desc = process failed to exit after 15 seconds"
2019-05-21T10:01:35.414+0100 [ERROR] client.alloc_runner.task_runner: failed to kill task: alloc_id=06db34f1-a57e-9004-27eb-942f9bf9e479 task=sigtest backoff=1m20s error="executor Shutdown failed: rpc error: code = Unknown desc = process failed to exit after 15 seconds"
2019-05-21T10:03:15.418+0100 [ERROR] client.alloc_runner.task_runner: failed to kill task: alloc_id=06db34f1-a57e-9004-27eb-942f9bf9e479 task=sigtest backoff=2m0s error="executor Shutdown failed: rpc error: code = Unknown desc = process failed to exit after 15 seconds"
2019-05-21T10:05:35.423+0100 [ERROR] client.alloc_runner.task_runner: failed to kill task: alloc_id=06db34f1-a57e-9004-27eb-942f9bf9e479 task=sigtest backoff=2m0s error="executor Shutdown failed: rpc error: code = Unknown desc = process failed to exit after 15 seconds"
2019-05-21T10:07:35.423+0100 [ERROR] client.alloc_runner.task_runner: failed to kill task. Resources may have been leaked: alloc_id=06db34f1-a57e-9004-27eb-942f9bf9e479 task=sigtest error="executor Shutdown failed: rpc error: code = Unknown desc = process failed to exit after 15 seconds"
```
By running strace against the trap.sh process I can see that the SIGINT is being sent correctly so Nomad is certainly still capable of sending signals:
```
wait4(-1, 0x7ffcfa3f96b0, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGINT {si_signo=SIGINT, si_code=SI_USER, si_pid=72333, si_uid=0} ---
rt_sigreturn()                          = -1 EINTR (Interrupted system call)
wait4(-1, 0x7ffcfa3f96b0, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
wait4(-1, 0x7ffcfa3f96b0, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGINT {si_signo=SIGINT, si_code=SI_USER, si_pid=72333, si_uid=0} ---
rt_sigreturn()                          = -1 EINTR (Interrupted system call)
wait4(-1,
```
This is not perfectly reproducible but appears to happen fairly consistently if the allocation has been running for at least twelve hours to a day. If the job is redeployed quickly then the SIGKILL correctly gets sent after five seconds.
```
wait4(-1, 0x7ffea53306c0, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGINT {si_signo=SIGINT, si_code=SI_USER, si_pid=18014, si_uid=0} ---
rt_sigreturn()                          = -1 EINTR (Interrupted system call)
wait4(-1, 0x7ffea53306c0, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
+++ killed by SIGKILL +++
```

--- a/drivers/exec/driver_test.go
+++ b/drivers/exec/driver_test.go
@@ -434,8 +434,8 @@ func TestExecDriver_HandlerExec(t *testing.T) {
 		if line == "" {
 			continue
 		}
-		// Skip systemd and rdma cgroups; rdma was added in most recent kernels and libcontainer/docker
-		// don't isolate them by default.
+		// Skip rdma subsystem; rdma was added in most recent kernels and libcontainer/docker
+		// don't isolate it by default.
 		if strings.Contains(line, ":rdma:") {
 			continue
 		}

--- a/drivers/exec/driver_test.go
+++ b/drivers/exec/driver_test.go
@@ -436,7 +436,7 @@ func TestExecDriver_HandlerExec(t *testing.T) {
 		}
 		// Skip systemd and rdma cgroups; rdma was added in most recent kernels and libcontainer/docker
 		// don't isolate them by default.
-		if strings.HasPrefix(line, "1:name=systemd") || strings.Contains(line, ":rdma:") {
+		if strings.Contains(line, ":rdma:") {
 			continue
 		}
 		if !strings.Contains(line, ":/nomad/") {

--- a/drivers/shared/executor/executor_linux.go
+++ b/drivers/shared/executor/executor_linux.go
@@ -684,7 +684,7 @@ func configureCgroups(cfg *lconfigs.Config, command *ExecCommand) error {
 	}
 
 	id := uuid.Generate()
-	cfg.Cgroups.Path = filepath.Join(defaultCgroupParent, id)
+	cfg.Cgroups.Path = filepath.Join("/", defaultCgroupParent, id)
 
 	if command.Resources == nil || command.Resources.NomadResources == nil {
 		return nil

--- a/drivers/shared/executor/executor_linux_test.go
+++ b/drivers/shared/executor/executor_linux_test.go
@@ -164,6 +164,59 @@ passwd`
 	}, func(err error) { t.Error(err) })
 }
 
+// TestExecutor_CgroupPaths asserts that process starts with independent cgroups
+// hierarchy created for this process
+func TestExecutor_CgroupPaths(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+	testutil.ExecCompatible(t)
+
+	testExecCmd := testExecutorCommandWithChroot(t)
+	execCmd, allocDir := testExecCmd.command, testExecCmd.allocDir
+	execCmd.Cmd = "/bin/bash"
+	execCmd.Args = []string{"-c", "sleep 0.2; cat /proc/self/cgroup"}
+	defer allocDir.Destroy()
+
+	execCmd.ResourceLimits = true
+
+	executor := NewExecutorWithIsolation(testlog.HCLogger(t))
+	defer executor.Shutdown("SIGKILL", 0)
+
+	ps, err := executor.Launch(execCmd)
+	require.NoError(err)
+	require.NotZero(ps.Pid)
+
+	state, err := executor.Wait(context.Background())
+	require.NoError(err)
+	require.Zero(state.ExitCode)
+
+	tu.WaitForResult(func() (bool, error) {
+		output := strings.TrimSpace(testExecCmd.stdout.String())
+		// sanity check that we got some cgroups
+		if !strings.Contains(output, ":devices:") {
+			return false, fmt.Errorf("was expected cgroup files but found:\n%v", output)
+		}
+		lines := strings.Split(output, "\n")
+		for _, line := range lines {
+			// Every cgroup entry should be /nomad/$ALLOC_ID
+			if line == "" {
+				continue
+			}
+
+			// Skip systemd and rdma cgroups; rdma was added in most recent kernels and libcontainer/docker
+			// don't isolate them by default.
+			if strings.Contains(line, ":rdma:") {
+				continue
+			}
+
+			if !strings.Contains(line, ":/nomad/") {
+				return false, fmt.Errorf("Not a member of the alloc's cgroup: expected=...:/nomad/... -- found=%q", line)
+			}
+		}
+		return true, nil
+	}, func(err error) { t.Error(err) })
+}
+
 func TestUniversalExecutor_LookupTaskBin(t *testing.T) {
 	t.Parallel()
 	require := require.New(t)

--- a/drivers/shared/executor/executor_linux_test.go
+++ b/drivers/shared/executor/executor_linux_test.go
@@ -203,8 +203,8 @@ func TestExecutor_CgroupPaths(t *testing.T) {
 				continue
 			}
 
-			// Skip systemd and rdma cgroups; rdma was added in most recent kernels and libcontainer/docker
-			// don't isolate them by default.
+			// Skip rdma subsystem; rdma was added in most recent kernels and libcontainer/docker
+			// don't isolate it by default.
 			if strings.Contains(line, ":rdma:") {
 				continue
 			}

